File: Financial_Coach_BackEnd/src/main/java/com/tp/financial_coach_backend/profile/entity/UserProfile.java
package com.tp.financial_coach_backend.profile.entity;

import com.tp.financial_coach_backend.auth.entity.User;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "user_profiles")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class UserProfile {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne
    @JoinColumn(name = "user_id", nullable = false, unique = true)
    private User user;

    // Informations personnelles
    @Column(name = "full_name")
    private String fullName;

    @Column(name = "age")
    private Integer age;

    @Column(name = "occupation")
    private String occupation;

    @Column(name = "country")
    private String country;

    @Column(name = "city")
    private String city;

    // Informations financières
    @Column(name = "monthly_income")
    private Double monthlyIncome;

    @Column(name = "monthly_savings")
    private Double monthlySavings;

    @Column(name = "risk_tolerance")
    private String riskTolerance; // low, medium, high

    // Objectifs financiers (stockés en JSON)
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "user_financial_goals", joinColumns = @JoinColumn(name = "profile_id"))
    @Column(name = "goal")
    @Builder.Default
    private List<String> financialGoals = new ArrayList<>();

    // Catégories de budget (stockées en JSON)
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "user_budget_categories", joinColumns = @JoinColumn(name = "profile_id"))
    @MapKeyColumn(name = "category")
    @Column(name = "amount")
    @Builder.Default
    private Map<String, Double> budgetCategories = new HashMap<>();

    // Préférences
    @Column(name = "preferred_language")
    private String preferredLanguage;

    @Column(name = "currency")
    private String currency;

    // Métadonnées
    @Column(name = "has_completed_onboarding")
    @Builder.Default
    private Boolean hasCompletedOnboarding = false;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // Calculer le taux d'épargne
    public Double getSavingsRate() {
        if (monthlyIncome != null && monthlySavings != null && monthlyIncome > 0) {
            return (monthlySavings / monthlyIncome) * 100;
        }
        return null;
    }
}

File: Financial_Coach_FrontEnd/lib/models/user_profile.dart
class UserProfile {
  final String userId;
  String? fullName;
  String? email;
  int? age;
  String? occupation;
  String? country;
  String? city;
  double? monthlyIncome;
  double? monthlySavings;
  List<String> financialGoals;
  String? riskTolerance; // 'low', 'medium', 'high'
  Map<String, double> budgetCategories; // Category -> Amount
  String? preferredLanguage;
  String? currency;
  bool hasCompletedOnboarding;
  DateTime? createdAt;
  DateTime? updatedAt;

  UserProfile({
    required this.userId,
    this.fullName,
    this.email,
    this.age,
    this.occupation,
    this.country,
    this.city,
    this.monthlyIncome,
    this.monthlySavings,
    List<String>? financialGoals,
    this.riskTolerance,
    Map<String, double>? budgetCategories,
    this.preferredLanguage = 'fr',
    this.currency = 'EUR',
    this.hasCompletedOnboarding = false,
    this.createdAt,
    this.updatedAt,
  })  : financialGoals = financialGoals ?? [],
        budgetCategories = budgetCategories ?? {};

  // Conversion depuis JSON
  factory UserProfile.fromJson(Map<String, dynamic> json) {
    return UserProfile(
      userId: json['userId'] as String,
      fullName: json['fullName'] as String?,
      email: json['email'] as String?,
      age: json['age'] as int?,
      occupation: json['occupation'] as String?,
      country: json['country'] as String?,
      city: json['city'] as String?,
      monthlyIncome: (json['monthlyIncome'] as num?)?.toDouble(),
      monthlySavings: (json['monthlySavings'] as num?)?.toDouble(),
      financialGoals: (json['financialGoals'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      riskTolerance: json['riskTolerance'] as String?,
      budgetCategories: (json['budgetCategories'] as Map<String, dynamic>?)
              ?.map((key, value) => MapEntry(key, (value as num).toDouble())) ??
          {},
      preferredLanguage: json['preferredLanguage'] as String? ?? 'fr',
      currency: json['currency'] as String? ?? 'EUR',
      hasCompletedOnboarding: json['hasCompletedOnboarding'] as bool? ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  // Conversion vers JSON
  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'fullName': fullName,
      'email': email,
      'age': age,
      'occupation': occupation,
      'country': country,
      'city': city,
      'monthlyIncome': monthlyIncome,
      'monthlySavings': monthlySavings,
      'financialGoals': financialGoals,
      'riskTolerance': riskTolerance,
      'budgetCategories': budgetCategories,
      'preferredLanguage': preferredLanguage,
      'currency': currency,
      'hasCompletedOnboarding': hasCompletedOnboarding,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  // Copie avec modifications
  UserProfile copyWith({
    String? userId,
    String? fullName,
    String? email,
    int? age,
    String? occupation,
    String? country,
    String? city,
    double? monthlyIncome,
    double? monthlySavings,
    List<String>? financialGoals,
    String? riskTolerance,
    Map<String, double>? budgetCategories,
    String? preferredLanguage,
    String? currency,
    bool? hasCompletedOnboarding,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return UserProfile(
      userId: userId ?? this.userId,
      fullName: fullName ?? this.fullName,
      email: email ?? this.email,
      age: age ?? this.age,
      occupation: occupation ?? this.occupation,
      country: country ?? this.country,
      city: city ?? this.city,
      monthlyIncome: monthlyIncome ?? this.monthlyIncome,
      monthlySavings: monthlySavings ?? this.monthlySavings,
      financialGoals: financialGoals ?? this.financialGoals,
      riskTolerance: riskTolerance ?? this.riskTolerance,
      budgetCategories: budgetCategories ?? this.budgetCategories,
      preferredLanguage: preferredLanguage ?? this.preferredLanguage,
      currency: currency ?? this.currency,
      hasCompletedOnboarding:
          hasCompletedOnboarding ?? this.hasCompletedOnboarding,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  // Vérifier si le profil est complet
  bool get isProfileComplete {
    return fullName != null &&
        age != null &&
        monthlyIncome != null &&
        financialGoals.isNotEmpty &&
        riskTolerance != null;
  }

  // Calculer le taux d'épargne
  double? get savingsRate {
    if (monthlyIncome != null &&
        monthlySavings != null &&
        monthlyIncome! > 0) {
      return (monthlySavings! / monthlyIncome!) * 100;
    }
    return null;
  }
}
